### ISSUE TITLE
Consider relying on vim-dispatch

### ISSUE THREAD

**Issue by Diablo-D3** [scrooloose/syntastic#699]
tpope has a new vim plugin out for async jobs. Watch the introductory video.

http://vimeo.com/63116209

https://github.com/tpope/vim-dispatch#readme

**Comment by treyhunner** [comment]
@q0LoCo I tried your async branch and it seems to work with flake8 for Python but not with ESLint for JavaScript. I'd open an issue there but you don't have the issue tracker enabled on your repository.

**Comment by q0LoCo** [comment]
@treyhunner it's opened now.

**Comment by acornejo** [comment]
since syntactic relies on external program's to check syntax, what is the main hurdle in doing something like this (oversimplifying, but want to understand if this approach is fundamentally broken):

1. run syntax checker external command forking it to the background and redirecting the output to to a file.
2. use cursorhold to trigger a periodicvim  function which checks if the file (created in the above step) has changed, and if so it does whatever syntastic already is doing to parse the output and put it in the location list.

**Comment by s0undt3ch** [comment]
@q0LoCo your branch often fails to update the UI. I'm using powerline(if that matters) and mostly pylint as the lint checker...

**Comment by lcd047** [comment]
@acornejo I believe the thread above answers your point: it's possible in principle, but it's less simple than you seem to believe, and the result is likely to suck.  Please feel free to prove me wrong though. :)

**Comment by lcd047** [comment]
@s0undt3ch Can you please take this to the [appropriate place](https://github.com/q0LoCo/syntastic/issues)? Thank you!

**Comment by chemzqm** [comment]
I've tried to use other async check plugins, but turns out I miss the features of syntastic so much.
Finally I realized the problem is not with syntastic, the actual reason is the lint tool runs so slowly.
So I made a plugin to check javascript syntax error only https://github.com/chemzqm/jscheck
it uses notify functions from syntastic which are very well  written :+1:

**Comment by bendavis78** [comment]
Out of curiosity, could we not somehow use vim's client/server model? That is, I could run my main vim session as a server, run the syntastic check in a separate client, and when checking is done have the signs and errors sent to the server.

**Comment by idbrii** [comment]
@bendavis78: There's not much benefit to initiating the check from another instance of vim. Most of the work is done by the external checker. Vim does some work to process the checker's output (see `:help efm`), but I don't see how we could send that from one vim session to another.

However, using vim's client/server model could work but there are concerns about requiring it for syntastic. AsyncCommand uses vim's client/server model to do async tasks: Run the checker in the background and when it completes, use `--remote` to send a signal to vim (but there's only been a proof of concept syntastic integration: stgpetrovic/syntastic-async@b3e3a9b). vim-dispatch works similarly, but uses a different signal to notify vim of completion. (See the above conversation about SIGWINCH.) I think the trickiest part is figuring out how to feed the output file into syntastic (probably more complex than `:cgetfile`).

**Comment by mtglsk** [comment]
Just so you guys know, version 7.4.1274 of vim added job functionality.

https://groups.google.com/forum/#!topic/vim_dev/_SbMTGshzVc

**Comment by justinmk** [comment]
Yep. So now everyone can stop pretending that job control isn't useful :)

**Comment by lcd047** [comment]
@mtglsk The interface is still under development, please see `:h job-options`.  Either way, a number of things need to change in syntastic before it could take advantage of background jobs without having to rewrite the existing 202 checkers.  Neovim had a similar interface for a while, and it inspired [neomake](https://github.com/benekastah/neomake), with a different approach than syntastic (see also #1253).

The forthcoming job interface in Vim will be similar to the one in Neovim, but it probably won't be compatible to it, despite several people (including Neovim's author) trying to reconcile the two APIs.  At least that's my understanding of Bram's position, at the time of this writing.

@justinmk Working PRs are still welcome.

**Comment by chemzqm** [comment]
Neovim could parse result of async make command by using 'errorformat' option automatically, but I'm wondering vim will have this feature or not.

**Comment by q0LoCo** [comment]
There is job_status() function in new job api. Vim can use autocmd to parse result if there is no way to pass a callback function to job_start().
A checker may do somethings on result, the problem is that current buffer may not be the same one it just checked.

**Comment by lcd047** [comment]
> A checker may do somethings on result, the problem is that current buffer may not be the same one it just checked.

Yup, that's one of the problems.

**Comment by parshap** [comment]
Vim 8 is out with "native" support for async functions [via channels](http://vimhelp.appspot.com/channel.txt.html)! Does this make sense for Syntastic to use?

**Comment by lcd047** [comment]
Yes, async checking shall be implemented, eventually.

**Comment by shinzui** [comment]
@lcd047 is there a separate issue for that?

**Comment by lcd047** [comment]
@shinzui Not that I know of.

**Comment by lencioni** [comment]
In the meantime, it looks like validator does async syntax checking. https://github.com/maralla/validator.vim

**Comment by lcd047** [comment]
@lencioni Right.  And so does [YouCompleteMe](http://valloric.github.io/YouCompleteMe/).  And [vim-go](https://github.com/fatih/vim-go).  And [neomake](https://github.com/neomake/neomake).  And [LaTeX-Box](https://github.com/LaTeX-Box-Team/LaTeX-Box). And [pyflakes-vim](https://github.com/kevinw/pyflakes-vim).  And [ghcmod-vim](https://github.com/eagletmt/ghcmod-vim). 
And other tools mentioned in [related](https://github.com/scrooloose/syntastic/issues/1253) [threads](https://github.com/scrooloose/syntastic/pull/370).  And many others I can't remember right now.  [Yet](https://github.com/w0rp/ale) [others](https://github.com/LucHermitte/vim-build-tools-wrapper) are getting there.  All of these projects have features that syntastic doesn't.  I strongly encourage you to use them instead of syntastic.

**Comment by blueyed** [comment]
@lcd047 as the de-facto current maintainer of Neomake I've often thought that my time would have been better spent by working on an async feature for syntastic, rather than fixing issues there all the time..
By now we have a test suite, but there are still fundamental issues, and all the groundwork already did and still does, and features it has are not available there.

**Comment by justinmk** [comment]
@lcd047 What are your thoughts on decoupling syntastic's library of checkers (similar to how [snippets](https://github.com/honza/vim-snippets) are separated from Ultisnips), so that other checker "engines" can use them? That would 100x more valuable than making syntastic itself asynchronous. It seems inevitable that people are going to re-invent these types of plugins, but if a library of checkers existed it would almost certainly guarantee that the checkers themselves would no longer be re-invented.

By "decouple" I don't necessarily mean moving the checkers to a separate repo, but just structuring the checkers so that they are isolated.

**Comment by lcd047** [comment]
@blueyed You (and many people above) assume that, now that Vim has support for jobs, what prevents syntastic from doing async checks is that it doesn't call checkers asynchronously.  That's not the case.  That is, it's true, but it's the least of the problems to be solved.  What _actually_ prevents syntastic from doing async checks right now is that half of its functions assume that `bufnr('')` returns the number of the buffer currently being checked, and that `b:` variables are local to the same buffer.  That problem was kind of lurking in syntastic before I came to the picture, but it was cast in stone by a design mistake that I made when I allowed checkers to have per-buffer overrides.  That happened long before Neovim, and before the above discussions about async checks.

So the problem I'm facing right now is how to fix the internal API, preferably without breaking compatibility with all third-party checkers out there.  It's a design problem, and no, simply breaking backwards compatibility won't solve it. _shrug_

**Comment by lcd047** [comment]
@justinmk Are we talking philosophy or logistics?

Philosophically, I think that:

1. Syntastic has built-in checkers; this has been a mistake;
2. Syntastic is using `errorformat` and the `makeprg` model; this has been a mistake;
3. Syntastic is trying (and failing) to solve some unreasonable parsing problems; this has been a mistake.

Logistically, I think that:

1. You can't meaningfully talk about a "library" without defining an API.  Syntastic did try to define such an API.  It failed, in that there are at least four methods of producing `loclist`s, and they are only superficially integrated:
    - `makeprg` / `lgetexpr` (most checkers)
    - the `preprocess` machinery (used f.i. by `pydocstyle`)
    - the `syntastic#c#GetLocList` machinery (used by most "real" compilers, such as `gcc`)
    - generating `loclists` from the output of other plugins (used by `vimlint`).
Integrating all these in a library would involve a major rewrite.
2. It won't do you much good to have "decoupled" checkers, simply because syntastic goes way beyond returning a list of errors.  It also compensates for a myriad of tiny flaws in Vim, and in the design of the 200+ supported linters (despite trying hard to limit those workarounds to a minimum).  Some time ago I browsed through [neomake's issues](https://github.com/neomake/neomake/issues), and as far as I can tell this is what its authors failed to take into account. _shrug_
3. You _could_ `grep` the code for `errorformat`s and put them into a library.  Some people take the view that syntastic is [useless](http://vi.stackexchange.com/questions/8100/how-to-switch-between-syntastic-error-window-and-editor-window#comment-13528), and everybody (especially new users) should learn to use the compiler features builtin to Vim.  I presume those users would find such a library useful, or at least an useful start.  Syntastic comes from there, and it's heading to the opposite direction.
4. VimL is a crappy language.  Among other things, it doesn't have named captures for regexps.  This could be overcome by relying on `python` and / or `python3` for parsing errors.  That would be a worthwhile project.  It would still not result in a library; just a major, very nice overhaul of three of the four methods mentioned at 1.
5. The bottleneck for implementing all coupling, decoupling, async checks, and whatever other features you guys are asking for has a name: your truly.  I'm the only maintainer left (for whatever reasons), and these days I can afford to dedicate only a very small amount of resources to syntastic.  I expect to be able to afford even less in the near future, despite the pressure you guys are building here.  I encourage you to fork the project and change it to fit your needs.  I'm an old fart, and I have my own priorities in the real life.  Don't wait for me.

**Comment by lencioni** [comment]
> despite the pressure you guys are building here

@lcd047 I didn't intend to build any pressure for you--simply to point people who are looking for this to something that might currently fit their needs. I rely on syntastic every day and I really appreciate all of your hard work that has gone into it.

> I'm the only maintainer left

I wonder if there are any folks in this thread who would be interested in helping out more?

**Comment by blueyed** [comment]
@lcd047 
> it was cast in stone by a design mistake that I made when I allowed checkers to have per-buffer overrides

Neomake has those, too: the basic idea there is that the bufnr and winnr (preferably via the new `win_getid` function family, but otherwise using a window-local var) will be remembered for the job that is being run.  And then e.g. location list results will only be opened when you are in the matching window/buffer (in case you move around while the job runs).
Any settings that are relevant should be retrieved when the job starts, or from the corresponding buffer later.

**Comment by justinmk** [comment]
@lcd047 Thanks for that valuable insight. Personally my interest is (was) to avoid redundant effort in the Vim ecosystem, but based on your answer here and at SO the situation appears grim.

> Some people take the view that syntastic is useless, and everybody (especially new users) should learn to use the compiler features builtin to Vim.

The reason IDEs work is because they ship with the entire stack. Plugins like [quickrun](https://github.com/thinca/vim-quickrun) and syntastic attempt to use tools found in arbitrary environments, configured (or not) by clueless users, to provide a light-weight IDE. And that's a recipe for endless bugs.

My preference would be to use the "compiler" feature of Vim. But it is incomplete, as you noted...

**Comment by lcd047** [comment]
> I didn't intend to build any pressure for you--simply to point people who are looking for this to something that might currently fit their needs.

@lencioni  So you started to make a list at the ~50th message in this thread.  How inspiring.

Well, now that you mention it, a list with other linter projects _could_ be pretty useful.  I'd suggest you create a [wiki page](https://github.com/scrooloose/syntastic/wiki/_new) for them and start collecting names, links, and basic characteristics (perhaps languages supported, async or not, etc.).  If / when this list starts getting substance I'll merge it into syntastic, to save GitHub the [trouble](https://github.com/scrooloose/syntastic/issues/1729) of rendering it.

> Any settings that are relevant should be retrieved when the job starts, or from the corresponding buffer later.

@blueyed Right, which is why `bufnr('')` references are fine in checkers and notifiers.  They're less than great in the [checker API](https://github.com/scrooloose/syntastic/blob/2b84b14cd1ef0d4bb74d91ca08a8818d84d8982d/plugin/syntastic/checker.vim#L72-L74), in code that runs [after the fact](https://github.com/scrooloose/syntastic/blob/2b84b14cd1ef0d4bb74d91ca08a8818d84d8982d/plugin/syntastic/checker.vim), in most of the [loclist API](https://github.com/scrooloose/syntastic/blob/2b84b14cd1ef0d4bb74d91ca08a8818d84d8982d/plugin/syntastic/loclist.vim), and in a few other `autocmd`-related places.  Since Vim 7.4.1640 they also [boobytrap](https://github.com/scrooloose/syntastic/issues/1127#issuecomment-216063694) `check_on_open`. _shrug_

> My preference would be to use the "compiler" feature of Vim. But it is incomplete, as you noted...

The `errorformat` plus `makeprg` model is flawed, not incomplete.  `makeprg` is broken because it makes it impossible to quote filenames robustly.  Syntastic found that out the hard way, and it switched to `system()` plus `lgetexpr`.  Running `makeprg` is also boobytrapped by `QuickFixCmdPre` and `QuickFixCmdPost` `autocmd`s (meaning, you'll end up calling code from other plugins if you aren't careful).

`errorformat` is flawed because:

1. it doesn't expose the parsing loop to the user
2. it doesn't allow adding extra fields to the standard line and column numbers
3. it depends on `isfname` in a way that makes it impossible to match certain combinations of filenames and errors
4. the interaction between parsing rules is less than useful, and
5. it has arbitrary (undocumented) limitations for no good reason.

`makeprg` and `errorformat` together are also flawed because they are essentially impossible to debug.  If there is anything that syntastic has done right, it has been logging.  It isn't perfect, because you can still run into `errorformat` problems.  But at least it can show you Vim's idea of what a linter should be doing.

Now, if you can live with these limitations then good for you.  They're certainly fine for many simple cases.  Some of us however have to make the best out of some hairy situations.

**Comment by lencioni** [comment]
> So you started to make a list at the ~50th message in this thread. How inspiring.

@lcd047 I'm not really sure I understand what you are referring to about starting a list here. I just learned about this other project, thought "I bet the fine folks who work on syntastic are working on async too", found this issue, noticed folks talking about async, and thought they might find the thing I just learned about to be useful. I'm not trying to make any lists or inspire anyone, I just wanted to help out and share some info. 😕 I apologize if I've upset you--that was certainly not my intention.

**Comment by bounceme** [comment]
I think any attempt to make on-the-fly checking would be a bad move, rather than having the neomake-like bufwrite execution. I saw a few different versions for on-the-fly(piping, making and writing and switching between temp buffers) and they all look flawed. vim doesn't seem to have a way to make a in memory pseudo-file for processing the errors.

**Comment by lcd047** [comment]
Here's an update for you guys.  The plan goes something like this:

* I'll make a 3.8 release, provided that no new bugs are reported the next few days.  This shall be the stable branch.  If new bugs are reported after the release I'll fix them, but I won't be adding any new features (or new checkers, for that matter).  This may eventually evolve into a stable 3.9 if needed.
* Right now I'm fixing the API for the checkers.  Checkers will be able to use custom parsing functions instead of `errorformat`.  Right now that means moving all JSON parsers from `preprocess` to the new parser sections of each checker, but there will be new parsing functions based on `if_python` and `if_perl`.  This means some checkers will depend on Vim being compiled with Python or Perl.
* Most internal checkers (hopefully all) will be migrated to the new API, but external third-party checkers will stop working.  In particular checkers that rely on calling a Vim function instead of running a program (such as [tsuquyomi](https://github.com/Quramy/tsuquyomi) and [omnisharp](https://github.com/OmniSharp/omnisharp-vim)) will break.  I plan to allow old-style checkers to run synchronously (along with the async checkers), and I'll eventually prove some upgrade path for checkers running Vim functions, but neither would happen at the beginning.
* The async functions still require a lot of bookkeeping.  I'll upload a new branch named `async` when I have them actually doing something useful.  They'll run with Vim 8.0, but they'll require a more recent Vim snapshot to work _reliably_ (a few critical problems have been fixed in Vim in the mean time).  At first they won't run with Neovim.  There will be hooks for Neovim though, provided that somebody in the know is willing to fill them in.
* A number of defaults will change.  If you want to suggest changes to defaults to your favourite checkers, now is the time to do it.  When the new branch gets released the defaults will again get locked for the next few years. :smile: Of course, I'll keep a detailed log with all incompatible changes, and I'll eventually update the docs, write an upgrade guide, etc.
* There are a number of other problems to be addressed, how to handle the `IsAvailable()` functions.  This is important because some checkers (hello, `rubocop`) take forever and a day to return a version.  So version queries might need to be run asynchronously -- but doing so would turn the API for `IsAvailable()` into a pain in the rear.  Let's hope some useful approach pops up eventually.
* If you plan to write new checkers for syntastic after the 3.8 release you'll have to use to the new API, based on whatever docs are available at the time.  I don't plan to offer much hand holding until the API stops changing.  Sorry about that.  Please feel free to explore using other checker engines in the mean time.
* When the new API in the `async` branch stops changing I'll call the result syntastic 4.0, and merge it to `master`.  All major issues would have to be fixed for that to happen.  And you probably want to stay with the stable branch in the mean time.

That's the basic outline.  Of course, that's just a plan.  Plans do typically change.

**Comment by blueyed** [comment]
@lcd047 
Thanks for the update and plans for a bright future.

Have you thought about using Neomake as a base for the new Syntastic?
Syntastic could be kept in maintenance mode as-is, and instead its features would get ported over to Neomake.  Neomake has a quite decent testing framework already, and has support for both Neovim and Vim's async modes.

I'd rather not waste development and support effort.  So if you do not like going the Neomake route then I might just put Neomake (from my side) into maintenance mode and help you out.. :)
But please consider that Neomake had been created (not by me though) as a proof-of-concept to make Syntastic async.

**Comment by lcd047** [comment]
@blueyed 
> Thanks for the update and plans for a bright future.

Oh absolutely, maintaining a healthy dose of skepticism is always a good idea. :smile:

> Have you thought about using Neomake as a base for the new Syntastic?

I have.  I still think syntastic is a lot easier as a foundation for what I'm trying to do than Neomake would be.  The vast majority of it is also my code, so I'm already pretty familiar with it.

> Neomake has a quite decent testing framework already,

On a side note, syntastic doesn't have tests for a single reason: I'm not the owner of this repository, and activating Travi-CI for it would require the intervention of the initial author.  I contacted him in the past and, while he was always nice, I did get a feeling he wants to put all this behind him, and he wouldn't appreciate being bothered with anything related to Vim these days.  Activating Travis-CI is simple but not completely trivial if you haven't done it in a long time, it might take a few exchanges back and forth.

> and has support for both Neovim and Vim's async modes.

I'm not worrying about Neovim at this point.  With the design I'm going to use adding support for Neovim shall be either trivial, or very hard.  I'm reasonably confident it's going to be trivial rather than very hard.

> But please consider that Neomake had been created (not by me though) as a proof-of-concept to make Syntastic async.

As far as I can tell the point of Neomake was to create an asynchronous `:make` for Neovim, completely integrated with compiler sets and directly extending them.  I think the author of Neomake considered syntastic grossly over-engineered.  I don't think Neomake was eve meant to be an async syntastic.

Now, is syntastic grossly over-engineered?  I'm sure it is.  I'm a scientist that happens to know a few things about computers, not a programmer, and when I started all this I didn't know substantially more about Vim than the average user (despite having already used Vim for 15+ years at the time).  It probably shows.  However, there is a number of things in syntastic that are the way they are for a number of reasons.  The author of Neomake was not aware of these reasons (which is by no means surprising, since `errorformat` and friends are a mine field), and that also shows.  I suppose people these days address this problem by writing blog posts about their projects, where they basically keep chronicles about their reasons and their designs.  But I'm an old fart, and blogs are way past my time. :smile:

**Comment by mgedmin** [comment]
> On a side note, syntastic doesn't have tests for a single reason: I'm not the owner of this repository, and activating Travi-CI for it would require the intervention of the initial author. I contacted him in the past and, while he was always nice, I did get a feeling he wants to put all this behind him, and he wouldn't appreciate being bothered with anything related to Vim these days.

It might be best to create a GitHub organization and ask the original maintainer to transfer repository ownership to the new org.

GitHub redirects old URLs when you transfer ownership, so existing links (and, more importantly, git checkouts) will continue to work.

**Comment by lcd047** [comment]
@mgedmin I'm not particularly unhappy with the current arrangement, and there are more pressing issues at the moment than enabling Travis-CI.  I might consider it later though.

**Comment by scrooloose** [comment]
Hey guys, I have also been thinking it would be a good idea to move syntastic out into an org - for a while now. @lcd047: you have been the main maintainer for a couple of years now and it is wrong for it be under my name after all this time. For one, I am getting credit for your hard work, and two, for the practical reasons you outlined above.

Thoughts?

One thing I would consider is putting the checkers and the core in different repos. Then you could have a very open policy with adding maintainers to the checker repo. For me, it was the tedium of maintaining the checkers that I burned out on. Perhaps that is a solution if you feel the same.

**Comment by lcd047** [comment]
@scrooloose You __do__deserve the credit, syntastic has been a very useful plugin for a years when I came around.  I just made it more maintainable, by making some optimisations, improving the docs, and by adding some obvious features.  But the overall design is yours, and it's beyond my league in creativity.

I'd be fine with either a GitHub community, or with taking over the project.  I'd also be fine with leaving it as it is, but yeah, there are a few practical problems that can only be solved by the owner of the project.

As for splitting the checkers from the core, that's probably a good idea.  It would make sense to split them by groups of languages, but that would leave the less popular languages in no man's land.  In fact the long tail is probably more important than the main languages, since languages like C/C++, Python, and Go already have plugins that can do much better than syntastic, simply by being more specialised.

**Comment by scrooloose** [comment]
OK, well that was scary but seems to have worked fine.

@lcd047 I have created the vim-syntastic org and added you as an owner (i.e. the highest permissions). I have then transferred syntastic into this org. All the links are redirecting fine (AFAICS) so it looks like a win.

Thanks for the kind words :-) I wouldn't undersell your hard work and persistence though. I think the syntastic project was very lucky that you came along just as I was running out of steam.

**Comment by lcd047** [comment]
@scrooloose Thank you, it seems to be working fine for me too.

**Comment by unphased** [comment]
Someone mentioned instability concerns; I use YCM and it already leaves `tern` stomping around 100% CPU semi-regularly. 

I really would prefer syntastic to stop blocking the UI because I have tmux binds that need Vim to stay responsive for the keybinds to stay responsive. 

Apologies for not reading the thread to the letter as that'd take the better half of a day which I'd be happy to devote if I could afford that. It sounds like architectural changes that will make syntastic even more amazing than it already is are in the pipes and they'll include robust async checkers. 👍

**Comment by mgedmin** [comment]
[ALE](https://github.com/w0rp/ale) is a Syntastic-like Vim plugin that works asynchronously.  AFAICT it supports fewer languages than Syntastic, but you could check it out today instead of waiting for Syntastic to get support for async.

**Comment by JohanBjoerklund** [comment]
Neomake is also an alternative

**Comment by lcd047** [comment]
@unphased Sorry to disappoint you, but for the last few months I've been struggling with health problems.  I'm a sick, old fart, and my priorities are those of a sick, old fart.  Support for async will be ready when / if I can get myself together enough to finish it.  Until then, you shouldn't really expect more than basic support for the existing code.  In the mean time you could always:

1. submit a patch to add support for async,
2. fork syntastic, or
3. use something else.

Not sure what else I could tell you. _shrug_

**Comment by gilbertwyw** [comment]
@lcd047 Hope you get well soon, i'm grateful for your great work on syntastic. 🙏

**Comment by unphased** [comment]
@lcd047 Absolutely, your contributions to the Vim ecosystem have NOT gone unnoticed!

**Comment by bradleyfalzon** [comment]
I know +1s are frowned upon, but I also wanted to say how much I loved and used Syntastic, it was one of the first plugins I used with vim, and we used it extensively. Love your work, and thank you.

**Comment by willsALMANJ** [comment]
Thank you, @lcd047 for all your work on Syntastic. I hope you feel better soon. If working on Syntastic adds stress to your life and asynchronous support will be a lot of work, you might want to try out ALE that @mgedmin suggested. I have been testing it recently and it works well -- for the languages I use it is like a drop-in replacement for Syntastic except it doesn't hang my terminal when I save. I worry that by the time Syntastic has asynchronous support that most of the userbase will have migrated to something else like it.

**Comment by bounceme** [comment]
https://github.com/vim-syntastic/syntastic/issues/699#issuecomment-285350988
there have already been several async linter runners made. ALE is one of the newest and will have the most changes and activity, sure, but tons of people have used syntastic for a long time and there is a huge amount of linter support here.

> And, last but not least, I have exactly zero use for all this, I'm pretty happy with the sync checks. shrug

**Comment by ErikBjare** [comment]
I just switched from syntastic to ALE which went pretty easy. I figured someone might wonder how to do that in a reasonable manner, so here's a basic port of my config: https://github.com/ErikBjare/dotfiles/commit/2b946b5d0f95ba056ce99bf0487c3ff414ce3c51

**Comment by vitaly-zdanevich** [comment]
Still blocking only, with waiting a few seconds after file saving? :(
